pbs_nonprint_characters.py: remove dependence on curses and make tests generic.

vccardenas opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Bug/feature Description
* Independence from curses module and make tests apply to any supported OS/platform.

#### Affected Platform(s)
* Linux

#### Cause / Analysis / Design
* The curses module might not build on recent platforms such as SLES 15 due to python bug.
* For maintainability the detection of certain OS/platforms was taken out.

#### Solution Description
* Use ANSI escape codes for terminal control instead of using curses module.
* Find out what the correct format for exporting shell functions and use that in the tests.

#### Testing logs/output

[TestNonprintingCharacters_nocurses_td-08.txt](https://github.com/PBSPro/pbspro/files/2823808/TestNonprintingCharacters_nocurses_td-08.txt)
[TestNonprintingCharacters_nocurses_td-09.txt](https://github.com/PBSPro/pbspro/files/2823809/TestNonprintingCharacters_nocurses_td-09.txt)

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

anamikau commented on test/tests/functional/pbs_nonprint_characters.py:
```diff
@@ -295,7 +292,7 @@ def test_nonprint_shell_function(self):
             # Adjustments in bash due to ShellShock malware fix in various OS
-            if self.osv in self.osv1 and self.pver >= self.osv1[self.osv]:
+            if self.is_osv1:
                 os.environ['BASH_FUNC_foo()'] = func
```
you have the value (BASH_FUNC_foo()) in self.n. can you not use that directly? Using that will get rid of the conditions.

vccardenas replied:
The self.n can be None and can't be used directly when exporting via os.environ.

vccardenas replied:
@anamikau , I initialize self.n for 'foo' and this will eliminate the need for the if else clause.

anamikau commented on test/tests/functional/pbs_nonprint_characters.py:
```diff
@@ -947,18 +944,22 @@ def test_terminal_control_hook_name(self):
-            self.server.manager(MGR_CMD_DELETE, HOOK, id=hook_name)
-            rv = self.server.create_hook(hook_name, a)
-        self.assertTrue(rv)
+            cmd = ['qmgr', '-c', 'delete hook %s' % hook_name]
```
I don't understand. why would the hook creation fail at the first time?

vccardenas replied:
If a hook with the same name already exists then hook creation will fail.

anamikau replied:
agreed. So you want to avoid using same names for hooks in the same test or move this to the setUp (deletion of hook part in case PTL setUp has failed or hook has not properly cleaned up by mom for some reason).

vccardenas replied:
I have the delete hook there in the very remote chance that the same hook name already existed.  The tests use different names for the hooks.

anamikau replied:
So I looked at the whole test and I still don't like the way this is done. how about checking if the hook_name exists, if yes then only delete it else go ahead and create them.

vccardenas replied:
Done

anamikau commented on test/tests/functional/pbs_nonprint_characters.py:
```diff
@@ -947,18 +944,22 @@ def test_terminal_control_hook_name(self):
         self.check_print_list_hook(hook_name, hook_name_esc)
         # Delete the hook
-        self.server.manager(MGR_CMD_DELETE, HOOK, id=hook_name, expect=True)
+        cmd = ['qmgr', '-c', 'delete hook %s' % hook_name]
```
why did you change it to use run_cmd?

vccardenas replied:
server.manager does not seem to work well with ANSI escape codes used for terminal control characters.

vccardenas commented on test/tests/functional/pbs_nonprint_characters.py:
```diff
@@ -69,43 +68,41 @@ def setUp(self):
-        else:
-            os_ver = ['', '']
-        return os_ver
+        # Check for ShellShock fix
```
Done

vccardenas commented on test/tests/functional/pbs_nonprint_characters.py:
```diff
@@ -69,43 +68,41 @@ def setUp(self):
+        fn = self.du.create_temp_file(body=foo_scr)
+        self.du.chmod(path=fn, mode=0755)
+        foo_msg = 'Failed to run foo_scr'
+        ret = self.du.run_cmd(self.server.hostname, cmd=fn)
```
This was meant to be run on server node, the test being a single node test.  The OS and bash version is likely the same on the server and mom nodes on systems like HPE and Cray anyway.

anamikau commented on test/tests/functional/pbs_nonprint_characters.py:
```diff
@@ -293,31 +289,13 @@ def test_nonprint_shell_function(self):
+                chk_var = self.n + '=() { a=%s; echo XX${a}YY}; }' % ch
+            out = self.n + \
+                '=() {  a=%s;\n echo XX${a}YY}\n}\nXX%sYY}' % (ch, ch)
+            script = ['#PBS -V']
```
looks like this is added twice

vccardenas replied:
Removed duplicate line.

anamikau commented on test/tests/functional/pbs_nonprint_characters.py:
```diff
@@ -947,18 +944,22 @@ def test_terminal_control_hook_name(self):
-        try:
-            rv = self.server.create_hook(hook_name, a)
-        except PbsManagerError:
+        cmd = ['qmgr', '-c', 'create hook %s' % hook_name]
```
create cmd1 for creating hook and cmd2 for deleting and use it at all places instead of over-writing cmd all the time.

vccardenas replied:
Done.

vccardenas commented:
I have added the recent test execution results to the previous ones above.

vccardenas commented:
@anamikau and @bayucan, thanks for the reviews and sign offs.
I have rebased and squashed my commits.
@maintainers, the changes are now ready to be merged.

vccardenas commented:
@mike0042, thanks for merging.
